frame/staking/src/lib.rs

```diff
@@ -1270,6 +1453,100 @@ impl<T: Trait> Module<T> {
 
 	// MUTABLES (DANGEROUS)
 
+	fn do_payout_nominator(who: T::AccountId, era: EraIndex, validators: Vec<(T::AccountId, u32)>)
+		-> DispatchResult
+	{
+		let mut nominator_ledger = <Ledger<T>>::get(&who).ok_or_else(|| Error::<T>::NotController)?;
+		if nominator_ledger.next_reward > era {
+			return Err(Error::<T>::InvalidEraToReward.into());
+		}
+
+		nominator_ledger.next_reward = era + 1;
+		<Ledger<T>>::insert(&who, &nominator_ledger);
+
+		let mut reward = Perbill::zero();
+		let era_reward_points = <ErasRewardPoints<T>>::get(&era);
+
+		// Note: iterator take only `MAX_NOMINATIONS` to avoid querying more validator exposure
+		// than necessary. Anyway a nominator can't validate more than `MAX_NOMINATIONS`.
+		for (validator, nominator_index) in validators.into_iter().take(MAX_NOMINATIONS) {
+			let commission = Self::eras_validator_prefs(&era, &validator).commission;
+			let validator_exposure = <ErasStakersClipped<T>>::get(&era, &validator);
+
+			if let Some(nominator_exposure) = validator_exposure.others
+				.get(nominator_index as usize)
+			{
+				if nominator_exposure.who != nominator_ledger.stash {
+					continue;
+				}
+
+				let nominator_exposure_part = Perbill::from_rational_approximation(
+					nominator_exposure.value,
+					validator_exposure.total,
+				);
+				let validator_point = era_reward_points.individual.get(&validator)
+					.map(|points| *points)
+					.unwrap_or_else(|| Zero::zero());
+				let validator_point_part = Perbill::from_rational_approximation(
+					validator_point,
+					era_reward_points.total,
+				);
+				reward = reward.saturating_add(
+					validator_point_part
+						.saturating_mul(Perbill::one().saturating_sub(commission))
+						.saturating_mul(nominator_exposure_part)
+				);
+			}
+		}
+
+		// Note: this is zero if the era is not finished yet.
+		let era_payout = <ErasValidatorReward<T>>::get(&era);
+		if let Some(imbalance) = Self::make_payout(&nominator_ledger.stash, reward * era_payout) {
+			Self::deposit_event(RawEvent::Reward(who, imbalance.peek()));
+		}
+
+		Ok(())
+	}
+
+	fn do_payout_validator(who: T::AccountId, era: EraIndex) -> DispatchResult {
+		let mut ledger = <Ledger<T>>::get(&who).ok_or_else(|| Error::<T>::NotController)?;
+		if ledger.next_reward > era {
+			return Err(Error::<T>::InvalidEraToReward.into());
+		}
+
+		ledger.next_reward = era + 1;
+		<Ledger<T>>::insert(&who, &ledger);
+
+		let era_reward_points = <ErasRewardPoints<T>>::get(&era);
+		let commission = Self::validators(&ledger.stash).commission;
```

Comment:
This is broken. Here we get the commission of the validator at the time of this call, not at the time of the era.

This means a validator can collect 84 eras worth of rewards, change their commission to 100%, and then take all the rewards, right?

Reply:
I see you created a new storage item, we just need to use it here.

Reply:
ah indeed I missed it we should take from eras_prefs same as in do_nominator_payout